vyper/venom/passes/dft.py

```diff
@@ -1,81 +1,152 @@
+import vyper.venom.effects as effects
 from vyper.utils import OrderedSet
+from vyper.venom.analysis.analysis import IRAnalysesCache
 from vyper.venom.analysis.dfg import DFGAnalysis
-from vyper.venom.basicblock import IRBasicBlock, IRInstruction, IRVariable
+from vyper.venom.analysis.liveness import LivenessAnalysis
+from vyper.venom.basicblock import IRBasicBlock, IRInstruction
 from vyper.venom.function import IRFunction
 from vyper.venom.passes.base_pass import IRPass
 
 
 class DFTPass(IRPass):
     function: IRFunction
-    inst_order: dict[IRInstruction, int]
-    inst_order_num: int
+    inst_offspring_count: dict[IRInstruction, int]
+    visited_instructions: OrderedSet[IRInstruction]
+    ida: dict[IRInstruction, OrderedSet[IRInstruction]]
 
-    def _process_instruction_r(self, bb: IRBasicBlock, inst: IRInstruction, offset: int = 0):
-        for op in inst.get_outputs():
-            assert isinstance(op, IRVariable), f"expected variable, got {op}"
-            uses = self.dfg.get_uses(op)
+    def __init__(self, analyses_cache: IRAnalysesCache, function: IRFunction):
+        super().__init__(analyses_cache, function)
+        self.inst_offspring_count = {}
 
-            for uses_this in uses:
-                if uses_this.parent != inst.parent or uses_this.fence_id != inst.fence_id:
-                    # don't reorder across basic block or fence boundaries
-                    continue
+    def run_pass(self) -> None:
+        self.visited_instructions: OrderedSet[IRInstruction] = OrderedSet()
+
+        self.dfg = self.analyses_cache.request_analysis(DFGAnalysis)
+        basic_blocks = list(self.function.get_basic_blocks())
+
+        self.function.clear_basic_blocks()
+        for bb in basic_blocks:
+            self._process_basic_block(bb)
+
+        self.analyses_cache.invalidate_analysis(LivenessAnalysis)
+
+    def _process_basic_block(self, bb: IRBasicBlock) -> None:
+        self.function.append_basic_block(bb)
+
+        self._calculate_dependency_graphs(bb)
+        self.instructions = list(bb.pseudo_instructions)
+
+        # Compute the number of instructions that are dependent on each instruction
+        self.visited_instructions = OrderedSet()
+        non_phi_instructions = list(bb.non_phi_instructions)
+        for inst in non_phi_instructions:
+            self._calculate_instruction_offspring_count_r(inst)
+
+        # Compute entry points in the graph of instruction dependencies
+        entry_instructions: OrderedSet[IRInstruction] = OrderedSet(non_phi_instructions)
```

Comment:
i think there is a performance error here. in the previous version, we would iterate into children with `get_uses` to bring instructions closer to their uses. but here we have no possibility to reorder the entry instructions.

Reply:
We can rearrange the entry instruction processing order. We are just not doing it just yet.